### ISSUE TITLE
Random weight init bug when using FSDP2 + CPU Offload

### ISSUE THREAD

**Issue by man2machine** [pytorch/torchtitan#3106]
### Bug description

When `fully_shard()` is applied with `CPUOffloadPolicy()` and the standard `to_empty(device='cpu')` + `init_weights()` pattern is used, sequential in-place random init calls (e.g. `nn.init.trunc_normal_`) on multiple fully sharded modules of the same shape produce bit-identical local shards. This applies in both cases:
  - across all ranks (every rank"s local shard has identical bytes), and
  - across all sequential init calls (every "trial" has identical local
    sum and first element).

Root cause: the local tensor of an FSDP2-wrapped `DTensor` under `CPUOffloadPolicy` is on CPU. `nn.init.trunc_normal_` -> `tensor.uniform_` dispatches to `DTensor`'s `__torch_dispatch__`, which routes through `OffsetBasedRNGTracker._distribute_region`. That context wraps the op in `torch.random.fork_rng(devices=[<cuda device>], device_type='cuda')`, which, per the docstring, "always" saves and restores CPU RNG state in addition to CUDA. The actual op runs on CPU and uses CPU RNG, but:
  1. Pre-yield: CPU RNG state is saved.
  2. yield -> CPU op runs, CPU RNG advances.
  3. finally: CPU RNG state is restored to the pre-yield value.

So CPU RNG never persistently advances across DTensor random ops, and since every rank starts with the same seed, every rank generates the same first-N elements of CPU RNG and stores them in its local shard. Sharding has no effect; sequential trials are bit-identical.

I found this bug when I was manually inspecting tensors on a training run with `cpu_offload=True`. The bad init cause the training loss to be much higher when using cpu offload, since each rank was effectively doing the same computation despite using different parameter shards.

Reproduction script:
https://gist.github.com/man2machine/9d33f81d745c1c3940c37a20fce17437

This is the output I get:
```
[rank 3] [vanilla] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 0] [vanilla] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 2] [vanilla] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 1] [vanilla] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 0] [vanilla]    trial=0  type=DTensor  local_numel=2048  local_sum=-0.3292308080158364  first=-0.0051163188181818
[rank 2] [vanilla]    trial=0  type=DTensor  local_numel=2048  local_sum=-1.4883650393703647  first=0.0291902758181095
[rank 0] [vanilla]    trial=1  type=DTensor  local_numel=2048  local_sum=0.3269380796514270  first=0.0265447869896889
[rank 2] [vanilla]    trial=1  type=DTensor  local_numel=2048  local_sum=0.1931311510871865  first=-0.0115052247419953
[rank 0] [vanilla]    trial=2  type=DTensor  local_numel=2048  local_sum=0.3211057182713830  first=-0.0112063353881240
[rank 2] [vanilla]    trial=2  type=DTensor  local_numel=2048  local_sum=1.2097331076762998  first=0.0083555206656456
[rank 0] [vanilla]    trial=3  type=DTensor  local_numel=2048  local_sum=-2.2544083424991186  first=-0.0200890358537436
[rank 2] [vanilla]    trial=3  type=DTensor  local_numel=2048  local_sum=0.3143082971178046  first=0.0422990322113037
[rank 3] [vanilla]    trial=0  type=DTensor  local_numel=2048  local_sum=-0.8609988162068021  first=-0.0138315781950951
[rank 1] [vanilla]    trial=0  type=DTensor  local_numel=2048  local_sum=0.8273666063278142  first=0.0168549474328756
[rank 3] [vanilla]    trial=1  type=DTensor  local_numel=2048  local_sum=0.0965054795151445  first=-0.0020712323021144
[rank 1] [vanilla]    trial=1  type=DTensor  local_numel=2048  local_sum=-0.2960697318158054  first=0.0330727398395538
[rank 3] [vanilla]    trial=2  type=DTensor  local_numel=2048  local_sum=0.1250745089755583  first=-0.0124917915090919
[rank 1] [vanilla]    trial=2  type=DTensor  local_numel=2048  local_sum=-1.3004788836005901  first=0.0298115238547325
[rank 3] [vanilla]    trial=3  type=DTensor  local_numel=2048  local_sum=0.7748959704185836  first=-0.0127454400062561
[rank 1] [vanilla]    trial=3  type=DTensor  local_numel=2048  local_sum=-0.1759570860849635  first=-0.0172545053064823
---
[rank 2] [fully_shard] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 0] [fully_shard] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 1] [fully_shard] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 3] [fully_shard] init_device=cpu mesh.size=4 shape=(4, 64, 32)
[rank 2] [fully_shard] trial=0  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 0] [fully_shard] trial=0  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 1] [fully_shard] trial=0  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 3] [fully_shard] trial=0  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 2] [fully_shard] trial=1  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 0] [fully_shard] trial=1  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 1] [fully_shard] trial=1  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 3] [fully_shard] trial=1  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 2] [fully_shard] trial=2  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 0] [fully_shard] trial=2  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 1] [fully_shard] trial=2  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 3] [fully_shard] trial=2  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 2] [fully_shard] trial=3  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 0] [fully_shard] trial=3  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 1] [fully_shard] trial=3  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
[rank 3] [fully_shard] trial=3  type=DTensor  local_numel=2048  local_sum=0.0296233969565947  first=-0.0001876695023384
```

### Versions

PyTorch 2.10+
